app/com/arpnetworking/metrics/portal/scheduling/JobExecutorActor.java

```diff
@@ -0,0 +1,180 @@
+     * @return The time until we should wake up. Non-negative.
+     */
+    /* package private */ static FiniteDuration timeUntilNextTick(final Instant now, final Instant timeToAwaken) {
+        final FiniteDuration delta = Duration.fromNanos(ChronoUnit.NANOS.between(now, timeToAwaken));
+        if (delta.lt(Duration.Zero())) {
+            return Duration.Zero();
+        } else if (delta.lt(TICK_INTERVAL)) {
+            return delta;
+        } else {
+            return TICK_INTERVAL;
+        }
+    }
+
+    private void execute(
+            final JobRepository<T> repo,
+            final Organization org,
+            final Job<T> job,
+            final Instant scheduled
+    ) {
+        job.execute(getSelf(), scheduled)
+                .handle((result, error) -> {
+                    if (error == null) {
+                        repo.jobSucceeded(job.getId(), org, scheduled, result);
+                    } else {
+                        repo.jobFailed(job.getId(), org, scheduled, error);
+                    }
+                    return null;
+                });
+        repo.jobStarted(job.getId(), org, scheduled);
+    }
+
+    @Override
+    public Receive createReceive() {
+        return receiveBuilder()
+                .match(Tick.class, message -> {
+                    _periodicMetrics.recordCounter("JobExecutorActor-ticks", 1);
+
+                    final JobRepository<T> repo = _jobRef.getRepository(_injector);
+                    final Optional<Job<T>> job = _jobRef.get(_injector);
+                    if (!job.isPresent()) {
+                        LOGGER.warn()
+                                .setMessage("no such job")
+                                .addData("jobRef", _jobRef)
+                                .log();
+                        getSelf().tell(PoisonPill.getInstance(), getSelf());
+                        return;
+                    }
+
+                    final UUID id = _jobRef.getJobId();
+                    final Organization org = _jobRef.getOrganization();
+                    final Optional<Instant> lastRun = repo.getLastRun(id, org);
+                    final Optional<Instant> nextRun = job.get().getSchedule().nextRun(lastRun);
+                    if (!nextRun.isPresent()) {
+                        LOGGER.info()
+                                .setMessage("job has no more scheduled runs")
+                                .addData("jobRef", _jobRef)
+                                .addData("schedule", job.get().getSchedule())
+                                .addData("lastRun", lastRun)
+                                .log();
+                        getSelf().tell(PoisonPill.getInstance(), getSelf());
+                        return;
+                    }
+
+                    final Instant now = _clock.instant();
+                    timers().startSingleTimer("TICK", Tick.INSTANCE, timeUntilNextTick(now, nextRun.get()));
+                    if (nextRun.get().isBefore(now)) {
```

Comment:
It's possible this schedules an immediate tick and an execution. This seems like it could result in a race condition because the tick would also schedule an execution unless the job had already executed and update itself in the repository. To me it seems that a tick should result in either an execution (which only when complete schedules a tick) or then another tick towards execution -- but not both.